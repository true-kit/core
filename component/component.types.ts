import {
	createElement,
} from 'react';
import {
	DescriptorWithMeta,
	DescriptorWithMetaMap,
	LikeFragment,
	LikeComponent,
	ToIntersect,
	CleanObject,
	Meta,
	GetMeta,
} from '../core.types';
import {
	DepsDescriptor,
	DepsMap,
	DepsMapBy,
} from '../deps';
import {
	Theme,
	ThemeRules,
} from '../theme/theme.types';

export interface ComponentDescriptor<
	N extends string,
	P extends object,
	DM extends DescriptorWithMetaMap,
> extends DescriptorWithMeta<N, P> {
	deps: DepsDescriptor<this, DM>;
	props: P;

	createComponent(render: ComponentRender<this>): LikeComponent<this['meta']>;
	createTheme<TS extends GetThemeSpec<P>>(rules: ThemeRules<TS>): Theme<TS>;
	createDeps(deps: Partial<DepsMapBy<DM>>): DepsMap
	createStrictDeps(deps: DepsMapBy<DM>): DepsMap;
}

type GetTheme<P> = P extends {theme?: infer P} ? P : never;
type GetThemeSpec<P> = P extends {theme?: Theme<infer TS>} ? TS : never;
type GetDeps<P> = P extends {deps?: infer D} ? D : never;

export type SlotProp<S extends SlotPropType> = null | (S extends (value: infer V) => infer R
	? SlotWithValue<S, V, R> | R
	: SlotWithoutValue<S> | S
) | Meta<S>

type SlotWithValue<S extends SlotPropType, V, R> = (parent: S, value: V) => R
type SlotWithoutValue<S extends SlotPropType> = (parent: S) => S

type SlotPropTypeInfer<P> = ToIntersect<P> extends Meta<infer S> ? S : never;

export type GetSlotsSpec<P extends object> = CleanObject<{
	[K in keyof P]-?: SlotPropTypeInfer<P[K]>;
}>

type SlotComponentProps<N, T> = (
	N extends 'children'
		? {name?: N}
		: {name: N}
) & (
	T extends (value: infer V) => any
		? {value: V; children?: T; } // <Slot name="..." value="...">{(value) => ...}</Slot>
		: {children?: T;} // <Slot name="...">...</Slot>
) & ({
	is?: SlotElement | [SlotElement | undefined, SlotElement | undefined];
})

export type SlotComponent<S extends object> = (
	(props: {
		[N in keyof S]: SlotComponentProps<N, S[N]>;
	}[keyof S]) => LikeFragment
);

export type ComponentRender<D extends ComponentDescriptor<any, any, any>> = (
	jsx: typeof createElement,
	props: D['meta'],
	env: {
		theme: GetTheme<D['meta']>;
		Slot: SlotComponent<GetSlotsSpec<D['meta']>>;
	},
	deps: GetDeps<D['meta']>,
) => LikeFragment;

type SlotValue = number | string | JSX.Element;

export type SlotContent = SlotValue | SlotValue[];
export type SlotPropType = SlotContent | ((value: object) => SlotContent);

export type SlotElement = null | {
	type: any;
	props: object;
}