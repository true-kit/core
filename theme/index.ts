export {
	getTheme,
	ThemeProvider,
	createThemeFor,
	composeThemes,
} from './theme';

export {
	Theme,
} from './theme.types';