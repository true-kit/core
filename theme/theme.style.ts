export class ThemeStyle {
	state = {};

	constructor(public name: string, public classes: object, private classNames: Function) {
		// console.log(name, classes, classNames.toString());
	}

	set(name: string, state: string | boolean) {
		this.state[name] = state;
		return this;
	}

	toString() {
		return this.classNames();
	}
}

ThemeStyle.prototype.constructor = String;