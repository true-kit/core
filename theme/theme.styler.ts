export class ThemeStyler {
	state = {};

	constructor(public name: string, public classes: object, private classNames: Function) {
		// console.log(name, classes, classNames.toString());
	}

	set(name: string, state: string | boolean) {
		this.state[name] = state;
		return this;
	}

	toDOMProps<T>(props: React.AllHTMLAttributes<T> = {}): React.AllHTMLAttributes<T> {
		props.className = (this as any) as string;
		return props;
	}

	toString() {
		return this.classNames();
	}
}

ThemeStyler.prototype.constructor = String;