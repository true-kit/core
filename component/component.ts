import { createElement, ReactElement, cloneElement } from 'react';
import { ComponentDescriptor, ComponentRender, SlotElement, SlotContent } from './component.types';
import { createDescriptor, getDescriptorOverride } from '../core';
import {
	createDepsDescriptor,
	createDepsBy,
	createStrictDepsBy,
} from '../deps';
import { DescriptorWithMetaMap } from '../core.types';
import { withEnvScope, getActiveEnvScope, getEnvContext } from '../env/env';
import { useTheme, createThemeFor } from '../theme';
import { EnvContextProps } from '../env/env.types';

export function createComponentDescriptor<
	N extends string,
	P extends object,
	DM extends DescriptorWithMetaMap,
>(
	name: N,
	props: P,
	deps?: DM,
): ComponentDescriptor<N, P, DM> {
	const $descriptor = createDescriptor(name).withMeta<P>() as ComponentDescriptor<N, P, DM>;
	const $deps = createDepsDescriptor($descriptor, Object(deps));

	$descriptor.deps = $deps;
	$descriptor.props = props;

	$descriptor.createComponent = (render) => createComponent($descriptor, render);
	$descriptor.createTheme = (rules) => createThemeFor($descriptor, rules);
	$descriptor.createDeps = (deps) => createDepsBy($deps, deps as any);
	$descriptor.createStrictDeps = (deps) => createStrictDepsBy($deps, deps as any);

	return $descriptor;
}

export function createComponent<
	D extends ComponentDescriptor<any, any, any>,
>(
	$descriptor: D,
	render: ComponentRender<D>,
): (props: D['meta']) => ReactElement {
	const $deps = $descriptor.deps;

	function Component(props: D['meta']) {
		const entry: EnvContextProps = {
			deps: null,
			theme: null,
			depsInjection: null,
			props,
		};

		return withEnvScope($descriptor, entry, () => {
			const ctx = getEnvContext();
			const overrides = ctx !== null && ctx.deps !== null ? ctx.deps.overrides : null;

			if (overrides !== null && overrides.has($descriptor)) {
				const override = getDescriptorOverride(overrides.get($descriptor)!, ctx!);

				if (override !== null && override.value !== Component) {
					return override.value(props);
				}
			}

			const theme = useTheme($descriptor, props, ctx);
			const deps = $deps.use(props, ctx);

			return render(
				createElement,
				props,
				{theme, Slot} as any,
				deps as any,
			);
		});
	}

	Component.$descriptor = $descriptor;
	Component.displayName = $descriptor.name;

	return Component;
}

type SlotProps = {
	name: string;
	value: any;
	children: React.ReactNode;
	is?: SlotElement | [SlotElement, SlotElement];
}

function Slot({name, children, value, is}: SlotProps) {
	if (name == null) {
		name = 'children';
	}

	const envScope = getActiveEnvScope();
	const scopeProps = (envScope !== null && envScope.ctx !== null) ? envScope.ctx.props : null;
	const content = scopeProps != null ? scopeProps[name] : undefined;

	if (is) {
		if (!is.hasOwnProperty('type')) {
			is = is[0] === null ? null : (is[0] || is[1] || null);
		}
	} else {
		is = null;
	}

	// Слот не нужен, так сказали свыше!
	if (content === null) {
		return null;
	}

	// Переопределение слота
	if (content !== undefined) {
		// Это перегрузка
		if (typeof content === 'function') {
			const result = content(children, value);

			// Слот не нужен
			if (result === null) {
				return null;
			} else if (result !== undefined) {
				return wrap(result, is);
			}
		} else {
			return wrap(content, is);
		}
	}

	// Значение есть, но оно именно null
	if (children === null) {
		return null;
	} else if (children === undefined) {
		return is; // возвращаем оборачиващий элемент
	}

	if (typeof children === 'function') {
		return wrap(children(value), is);
	}

	return wrap(children, is);
}

function wrap(content: any, is: any) {
	if (is !== null) {
		content = content && cloneElement(
			is as React.ReactElement,
			is.props,
			content,
		);
	}

	return content;
}
