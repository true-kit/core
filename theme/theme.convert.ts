import { ThemeRules, ThemeRule } from './theme.types';
import { IRuleDefinitions, CSSMap, css } from '@artifact-project/css';

export function createToStringCode(code: string[], name: string, classes: object) {
	if (typeof classes === 'string') {
		code.push(`
			if (n === "${name}") {
				return (c || '');
			}
		`);
		return;
	}

	const str = Object.entries(classes).map(([key, val]) => {
		if (key === '$root') {
			return name === 'host' ? `(c.$root || '')` : '';
		} else if (typeof val === 'object') {
			return `(k = '${key}', s[k] ? (c[k].$root ? ' ' + c[k].$root : '') + ' ' + (c[k][s[k]] || '') : '')`;
		} else {
			return `(k = '${key}', s[k] ? ' ' + (c[k] || '') : '')`;
		}
	}).filter(x => !!x).join(' + ');

	code.push(`
		if (n === "${name}") {
			return (${str ? str : '""'});
		}
	`);
}

const SEP = '--';
const COLON_CODE = ':'.charCodeAt(0);

type SelectorToken = {
	type: '+' | '.' | 'fn' | ':';
	raw: string;
}

class Selector {
	constructor(raw: SelectorToken['raw'], type: SelectorToken['type'] = '.', private tokens: SelectorToken[] = []) {
		this.tokens.push({raw, type});
	}

	add(raw: SelectorToken['raw'], type: SelectorToken['type'] = '.') {
		return new Selector(raw, type, this.tokens.slice(0));
	}

	toString() {
		let fnOpened = false;
		let prevType: SelectorToken['type'];

		return this.tokens.reduce((s, {raw, type}, idx) => {
			if (type === '.') {
				if (raw === '&') {
					s += '.' + this.tokens[0].raw;
				} else if (prevType !== '.' && idx > 0) {
					s += '.' + this.tokens[0].raw + SEP + raw;
				} else {
					s += (prevType === '.' ? SEP : '.') + raw;
				}
			} else if (type === ':') {
				s += raw;
			} else {
				if (fnOpened) {
					s += ')';
					fnOpened = false;
				}

				if (type === '+') {
					s += ' + ';
				} else if (type === 'fn') {
					s += raw + '('
					fnOpened = true;
				}
			}

			if (fnOpened && (this.tokens.length - idx === 1)) {
				s += ')';
				fnOpened = false;
			}

			prevType = type;
			return s;
		}, '');
	}
}

export function convertThemeRulesToCSSRules(rules: ThemeRules<any>, css = {}) {
	Object.entries(rules).forEach(([name, rule]) => {
		convertThemeRuleToCSSRules(new Selector(name), rule, css);
	});

	return css as IRuleDefinitions;
}

function convertThemeRuleToCSSRules(sel: Selector, rule: ThemeRule<any, any>, css: object) {
	Object.entries(rule).forEach(([key, value]) => {
		if (value && typeof value === 'object') {
			if (key.charCodeAt(0) === COLON_CODE) {
				if (key === ':modifiers' || key === ':self') {
					convertThemeRuleToCSSRules(sel, value, css);
				} else if (key === ':not') {
					convertThemeRuleToCSSRules(sel.add(key, 'fn'), value, css);
				} else {
					convertThemeRuleToCSSRules(sel.add(key, ':'), value, css);
				}
			} else if (key === '+') {
				convertThemeRuleToCSSRules(sel.add('+', '+'), value, css);
			} else if (key === '&') {
				convertThemeRuleToCSSRules(sel.add('&'), value, css);
			} else {
				convertThemeRuleToCSSRules(sel.add(key), value, css);
			}
		} else {
			const name = sel.toString();
			css[name] = css[name] || {};
			css[name][key] = value;
		}
	});
}

export function convertCSSRulesToCSSClasses<T extends IRuleDefinitions>(rules: T): CSSMap<T> {
	return css(rules)
}

export function convertCSSClassesToThemeClasses<T extends IRuleDefinitions, M extends CSSMap<T>>(map: M) {
	const classes = {
	};

	Object.entries(map).forEach(([key, val]) => {
		const [elem, mod, state] = key.split(SEP);

		if (classes[elem] == null) {
			classes[elem] = map[elem] ? {$root: map[elem]} : {};
		}

		if (mod != null && !state && classes[elem][mod] == null) {
			classes[elem][mod] = val;
		}

		if (state != null) {
			if (!classes[elem][mod] || typeof classes[elem][mod] !== 'object') {
				classes[elem][mod] = classes[elem][mod] ? {$root: classes[elem][mod]} : {};
			}

			classes[elem][mod][state] = val;
		}
	});

	!classes.hasOwnProperty('host') && (classes['host'] = {});
	!classes.hasOwnProperty('elements') && (classes['elements'] = {});

	return classes;
}
