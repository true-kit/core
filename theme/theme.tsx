import {
	ThemeRules,
	ThemeSpec,
	Theme,
	ThemeRegistry,
	LikeComponent,
	ThemeOverride,
	GetThemeFromDescriptor,
	ThemeOverrideIndex,
	AllThemes,
} from './theme.types';

import {
	convertThemeRulesToCSSRules,
	convertCSSRulesToCSSClasses,
	convertCSSClassesToThemeClasses,
	createToStringCode,
} from './theme.convert';

import { ThemeStyle } from './theme.style';

import { DescriptorWithMeta, Last, Cast } from '../core.types';
import { EnvContextEntry } from '../env/env.types';
import { createEnvContextProvider, getEnvContext, getActiveEnvScope } from '../env/env';


export function createThemeRegistryFor<
	D extends DescriptorWithMeta<any, any>
>(
	descriptor: D,
	themesMap: AllThemes<D>,
	overrides?: ThemeOverride[] | null,
): ThemeRegistry {
	return createThemeRegistry(
		Object.values(themesMap).map(theme => theme) as Theme<any>[],
		overrides,
	);
}

export function createThemeFor<
	D extends DescriptorWithMeta<any, {theme?: Theme<any>}>,
	TS extends ThemeSpec = NonNullable<D['meta']['theme']>['spec'],
>(descriptor: D, rules: ThemeRules<TS>): Theme<TS> {
	const cssRules = convertThemeRulesToCSSRules(rules);
	const cssClasses = convertCSSRulesToCSSClasses(cssRules);
	const classes = convertCSSClassesToThemeClasses(cssClasses);
	const toStringCode = [
		`
		var k;
		var n = this.name;
		var s = this.state;
		var c = this.classes;
		`
	];

	createToStringCode(toStringCode, 'host', classes['host']);
	Object.entries(classes['elements']).forEach(([name, classes]) => {
		createToStringCode(toStringCode, name, classes);
	});

	const classNames = Function(`
		${toStringCode.join('\n')}
		return '';
	`);

	function _for(name: string) {
		const store = this.store;

		if (store === null || store.hasOwnProperty(name)) {
			const style = new ThemeStyle(
				name,
				name === 'host' ? classes[name] : classes['elements'][name],
				classNames,
			) as any;

			if (store === null) {
				return style
			}

			store[name] = style;
		}

		return store[name];
	}

	return {
		descriptor: descriptor as any,
		for: _for,
		cssRules,
		classes,
		store: null,
		persist(store: object) {
			const wr = Object.create(this);
			wr.store = store;
			return wr;
		},
	} as Theme<TS>;
}

const nullTheme = createThemeFor(null as any, {
	host: {},
	elements: {},
});

export const ThemeProvider = createEnvContextProvider('theme');

export function getTheme<
	D extends DescriptorWithMeta<any, {theme?: Theme<any>}>,
	T extends Theme<any> = NonNullable<D['meta']['theme']>,
>(descriptor: D, props: {theme?: T}): T {
	let theme = props.theme;

	if (theme == null) {
		theme = nullTheme as T;

		let ctx = getEnvContext();

		while (ctx && ctx.theme) {
			const {
				map,
				overrides,
			} = ctx.theme;

			if (
				overrides !== null
				&& getActiveEnvScope() !== null
				&& overrides.has(descriptor)
			) {
				const overTheme = getThemeOverride(overrides.get(descriptor)!, ctx);
				if (overTheme !== null) {
					theme = overTheme as T;
					break;
				}
			}

			if (map !== null && map.has(descriptor)) {
				theme = map.get(descriptor) as T;
				break;
			}

			ctx = ctx.parent!;
		}
	}

	return theme;
}

function getThemeOverride(list: ThemeOverride[], ctx: EnvContextEntry): Theme<any> | null {
	for (let i = 0; i < list.length; i++) {
		let {value: theme, xpath} = list[i];
		let scope = getActiveEnvScope();

		XPATH: for (let x = 0, xn = xpath.length; x < xn; x++) {
			const descr = xpath[x];

			while (true) {
				scope = scope!.parent;

				if (scope === null || scope.ctx === ctx) {
					theme = null as any;
					break XPATH;
				}

				if (scope.owner === descr) {
					break;
				}
			}
		}

		if (theme !== null) {
			return theme;
		}
	}

	return null;
}

export function createThemeRegistry(
	themes: Theme<any>[] | null,
	overrides?: ThemeOverride[] | null,
): ThemeRegistry {
	return {
		map: !themes ? null : themes.reduce((map, theme) => {
			map.set(theme.descriptor, theme);
			return map;
		}, new Map),

		overrides: !overrides ? null : overrides.reduce((index, override) => {
			const xpath = override.xpath.slice();
			const Node = xpath.pop();

			if (xpath.length >= 1 && Node) {
				index.set(Node, (index.get(Node) || []).concat({
					value: override.value,
					xpath,
				}));
			} else {
				console.warn('[@truekit/core: theme] Invalid override: ', override);
			}

			return index;
		}, new Map as ThemeOverrideIndex),
	};
}

export function createThemeOverrideFor<
	X extends DescriptorWithMeta<any, {theme?: Theme<any>}>[]
>(...xpath: X): (theme: GetThemeFromDescriptor<Last<X>>) => ThemeOverride {
	return (theme): ThemeOverride => ({
		value: theme as any,
		xpath,
	});
}