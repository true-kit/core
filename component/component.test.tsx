import * as React from 'react'; // @todo: Переехать на `jsx`
import * as ReactDOM from 'react-dom';

import { createComponentDescriptor } from './component';
import {
	Deps,
	DepsProvider,
	createDepsRegistry,
	createDeps,
	createDepOverride,
} from '../deps';
import { Theme } from '../theme';
import {
	createThemeRegistryFor,
	ThemeProvider,
	createThemeRegistry,
} from '../theme/theme';
import { resetCSS } from '@artifact-project/css';
import { SlotContent, SlotProp } from './component.types';

const $Icon = createComponentDescriptor('@truekit/core/cmp: Icon', {} as IconProps, {});
const $Btn = createComponentDescriptor('@truekit/core/cmp: Btn', {} as BtnProps, {
	Icon: $Icon,
});
const $Card = createComponentDescriptor('@truekit/core/cmp: Card', {} as CardProps, {});

type IconProps = {
	name: string;
	theme?: Theme<{
		host: {};
		elements: {};
	}>;
}

type BtnProps = {
	type: 'text' | 'password';
	value: string;
	theme?: Theme<{
		host: Pick<BtnProps, 'type'>;
		elements: {value: true};
	}>;
	deps?: Deps<typeof $Btn.deps>;
}

type CardProps = {
	children: SlotContent;
}

const Icon = $Icon.createComponent((jsx, {name}, {theme}) => {
	const hostTheme = theme.for('host');
	return jsx('i', hostTheme.toDOMProps(), <>{name}</>);
});

const EmIcon = $Icon.createComponent((jsx, {name}, {theme}) => {
	const hostTheme = theme.for('host');
	return jsx('em', hostTheme.toDOMProps(), <>{name}</>);
});

const StrongIcon = $Icon.createComponent((jsx, {name}, {theme}) => {
	const hostTheme = theme.for('host');
	return jsx('strong', hostTheme.toDOMProps(), <>{name}</>);
});

const SpanIcon = $Icon.createComponent((jsx, {name}, {theme}) => {
	const hostTheme = theme.for('host');
	return jsx('span', hostTheme.toDOMProps(), <>{name}</>);
});

const Btn = $Btn.createComponent((jsx, {type, value}, {theme}, {Icon}) => {
	const hostTheme = theme.for('host').set('type', type);

	return jsx('button', hostTheme.toDOMProps({type}), <>
		<Icon name="pic" />
		<span className={theme.for('value')}>{value}</span>
	</>);
});

const Card = $Card.createComponent((jsx, {children}, {}) => {
	return jsx('div', {}, children);
});

let btnTheme: NonNullable<BtnProps['theme']>;
let iconTheme: NonNullable<IconProps['theme']>;

beforeAll(() => {
	process.env.NODE_ENV = 'production';
	resetCSS(0);

	btnTheme = $Btn.createTheme({
		host: {color: 'black', ':modifiers': {type: {password: {color: 'red'}}}},
		elements: {value: {color: '#333'}},
	});

	iconTheme = $Icon.createTheme({
		host: {color: 'blue'},
		elements: {},
	});
});

describe('component', () => {
	it('Btn without theme and deps', () => {
		expect(render(<Btn type="password" value="foo" />)).toMatchSnapshot();
	});

	it('Btn with inline deps and without theme', () => {
		expect(render(<Btn
			deps={{Icon}}
			type="password"
			value="bar" />
		)).toMatchSnapshot();
	});

	it('Btn with inline deps and inline root theme', () => {
		expect(render(<Btn
			deps={{Icon}}
			theme={btnTheme}
			type="password"
			value="baz" />
		)).toMatchSnapshot();
	});

	it('Btn with all deps and themes', () => {
		const deps = $Btn.createStrictDeps({
			[$Icon.id]: Icon,
		});
		const allDeps = createDepsRegistry([deps]);
		const allThemes = createThemeRegistryFor($Btn, {
			[$Btn.id]: btnTheme,
			[$Icon.id]: iconTheme,
		});

		expect(render(
			<DepsProvider value={allDeps}>
				<ThemeProvider value={allThemes}>
					<Btn type="password" value="qux" />
				</ThemeProvider>
			</DepsProvider>
		)).toMatchSnapshot();
	});

	it('Btn with deps overrides', () => {
		const defaultDeps = createDeps({[$Icon.id]: Icon});
		const iconOver1 = createDepOverride($Icon, [$Card])(EmIcon);
		const iconOver2 = createDepOverride($Icon, [$Card], {name: 'over-2'})(StrongIcon);
		const iconOver3 = createDepOverride($Icon, [$Card, $Card])(SpanIcon);

		expect(render(
			<DepsProvider value={createDepsRegistry([defaultDeps], [iconOver1, iconOver2, iconOver3])}>
				<Icon name="default"/>{'\n'}
				<Card>{'\n\t'}
					<Icon name="over-1"/>{'\n\t'}
					<Card>{'\n\t\t'}
						<Icon name="over-2"/>{'\n\t\t'}
						<Icon name="over-3"/>{'\n\t'}
					</Card>{'\n'}
				</Card>
			</DepsProvider>
		)).toMatchSnapshot();
	});
});

describe('component slots', () => {
	type BlockProps = {
		children?: SlotProp<SlotContent>;
		theme?: Theme<{
			host: {},
			elements: {},
		}>;
		value?: string;
		data?: object;
		slotNum?: SlotProp<number>;
		slotStr?: SlotProp<string>;
		slotFragment?: SlotProp<SlotContent>;
		slotWithValue?: SlotProp<(value: {username:string}) => SlotContent>;
	}

	const $Block = createComponentDescriptor('@truekit/core/cmp: Block', {} as BlockProps);
	const Block = $Block.createComponent((jsx, {}, {Slot, theme}) => {
		return (
			<div className={theme.for('host')}>{'\n\t'}
				<Slot/>{'\n\t'}
				<div>num: <Slot name="slotNum">{123}</Slot></div>{'\n\t'}
				<div>val: <Slot name="slotStr">wow</Slot></div>{'\n\t'}
				<div>fragment: <Slot name="slotFragment"><b>bold</b>!</Slot></div>{'\n\t'}
				<div>{'\n\t\t'}
					callback:{'\n\t\t'}
					<Slot name="slotWithValue" value={{username: "RubaXa"}}>
						{({username}) => <h2>Hi, {username}!</h2>}
					</Slot>{'\n\t'}
				</div>{'\n'}
			</div>
		);
	});

	it('default slots', () => {
		expect(render(<Block/>)).toMatchSnapshot();
	});

	it('override num and str slot', () => {
		expect(render(<Block
			slotNum={321}
			slotStr={null}
		/>)).toMatchSnapshot();
	});

	it('override slot with parent', () => {
		expect(render(<Block
			slotNum={(parent) => parent * 2}
			slotWithValue={(parent, value) => <>{parent(value)}<br/>Yes, {value.username}.</>}
		/>)).toMatchSnapshot();
	});

	it('slots with nested', () => {
		const redTheme = $Block.createTheme({
			host: {color: 'red'},
			elements: {},
		});
		const greenTheme = $Block.createTheme({
			host: {color: 'green'},
			elements: {},
		});
		const blueTheme = $Block.createTheme({
			host: {color: 'blue'},
			elements: {},
		});

		expect(render(
			<ThemeProvider value={createThemeRegistry([blueTheme])}>
				<Block slotNum={(parent) => parent * 2} theme={redTheme}>{'\n\t'}
					<Block slotNum={(parent) => parent * 3} slotWithValue={null} theme={greenTheme}/>{'\n\t'}
					---{'\n\t'}
					<Block slotNum={(parent) => parent * 4} slotWithValue={null}/>{'\n\t'}
				</Block>
			</ThemeProvider>
		)).toMatchSnapshot();
	});
});

const root = document.createElement('div');

function render(fragment: JSX.Element) {
	ReactDOM.render(fragment, root);
	return root.innerHTML;
}
