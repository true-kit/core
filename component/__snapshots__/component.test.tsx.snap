// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Btn with all deps and themes 1`] = `"<button type=\\"password\\" class=\\"_0 _1\\"><i class=\\"_3\\">pic</i><span class=\\"_2\\">qux</span></button>"`;

exports[`component Btn with deps overrides 1`] = `
"<i class=\\"\\">default</i>
<div>
	<em class=\\"\\">over-1</em>
	<div>
		<strong class=\\"\\">over-2</strong>
		<span class=\\"\\">over-3</span>
	</div>
</div>"
`;

exports[`component Btn with inline deps and inline root theme 1`] = `"<button type=\\"password\\" class=\\"_0 _1\\"><i class=\\"\\">pic</i><span class=\\"_2\\">baz</span></button>"`;

exports[`component Btn with inline deps and without theme 1`] = `"<button type=\\"password\\" class=\\"\\"><i class=\\"\\">pic</i><span class=\\"\\">bar</span></button>"`;

exports[`component Btn without theme and deps 1`] = `
"<button type=\\"password\\" class=\\"\\"><i>
		Dep '@truekit/core/cmp: Icon' (aka 'Icon') not found
	</i><span class=\\"\\">foo</span></button>"
`;

exports[`component slots default slots 1`] = `
"<div class=\\"\\">
	
	<div>num: 123</div>
	<div>val: wow</div>
	<div>fragment: <b>bold</b>!</div>
	<div>
		callback:
		<h2>Hi, RubaXa!</h2>
	</div>
</div>"
`;

exports[`component slots override num and str slot 1`] = `
"<div class=\\"\\">
	
	<div>num: 321</div>
	<div>val: </div>
	<div>fragment: <b>bold</b>!</div>
	<div>
		callback:
		<h2>Hi, RubaXa!</h2>
	</div>
</div>"
`;

exports[`component slots override slot with parent 1`] = `
"<div class=\\"\\">
	
	<div>num: 246</div>
	<div>val: wow</div>
	<div>fragment: <b>bold</b>!</div>
	<div>
		callback:
		<h2>Hi, RubaXa!</h2><br>Yes, RubaXa.
	</div>
</div>"
`;

exports[`component slots slot with is defaults 1`] = `
"<h1>Primary</h1>
Sub
<h4></h4>
"
`;

exports[`component slots slot with is header is null 1`] = `
"
Sub
<h4></h4>
"
`;

exports[`component slots slot with is header is undefined 1`] = `
"<h1>Primary</h1>
Sub
<h4></h4>
"
`;

exports[`component slots slot with is lastHeaderEl is h3 1`] = `
"<h1>Primary</h1>
Sub
<h3></h3>
"
`;

exports[`component slots slot with is lastHeaderEl is null 1`] = `
"<h1>Primary</h1>
Sub

"
`;

exports[`component slots slot with is lastHeaderEl is undefined 1`] = `
"<h1>Primary</h1>
Sub
<h4></h4>
"
`;

exports[`component slots slot with is subHeaderEl is h2 1`] = `
"<h1>Primary</h1>
<h2>Sub</h2>
<h4></h4>
"
`;

exports[`component slots slot with is subHeaderEl is h2 with override 1`] = `
"<h1>Primary</h1>
<h2>Wow!</h2>
<h4></h4>
"
`;

exports[`component slots slots with nested 1`] = `
"<div class=\\"_4\\">
	
	<div class=\\"_5\\">
	
	<div>num: 369</div>
	<div>val: wow</div>
	<div>fragment: <b>bold</b>!</div>
	<div>
		callback:
		
	</div>
</div>
	---
	<div class=\\"_6\\">
	
	<div>num: 492</div>
	<div>val: wow</div>
	<div>fragment: <b>bold</b>!</div>
	<div>
		callback:
		
	</div>
</div>
	
	<div>num: 246</div>
	<div>val: wow</div>
	<div>fragment: <b>bold</b>!</div>
	<div>
		callback:
		<h2>Hi, RubaXa!</h2>
	</div>
</div>"
`;
